### PR TITLE
[firebase_messaging] Handle background messages on iOS

### PR DESCRIPTION
**Pull request by tobiasjunsten**
## Description
Enable backgound handling on iOS to work the same way as android.

## Related Issues
See [Issue 47](https://github.com/FirebaseExtended/flutterfire/issues/47)

## Related pull request
This pull request was inspired by this one: [Pull request 53](https://github.com/FirebaseExtended/flutterfire/pull/53)

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by Finn0811** [comment]
Really looking forward to this, thanks for your work!

**Comment by shell32** [comment]
Thanks for your great work, only worked in background .when app killed not work?

**Comment by athornz** [comment]
@shell32 what device are you testing on? If you force kill an app on iOS (common when testing things like this) then iOS will not deliver push notifications to that app until the app or device is restarted.

**Comment by shell32** [comment]
> @shell32 what device are you testing on? If you force kill an app on iOS (common when testing things like this) then iOS will not deliver push notifications to that app until the app or device is restarted.

iphone 6 , ios 12

**Comment by tobiasjunsten** [comment]
@shell32, just as @athornz says the app is not receiving any push notifications if the app has been force-quit. If the app has been suspended by the system (which happens quite fast) the app should be woken up and get some time to process a notification. When I've been testing I've also noticed that it can take a couple of minutes sometimes before the app is woken up if it has been suspended. You can read a bit more here: [https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623013-application#discussion](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623013-application#discussion)

**Comment by gustavoergalves** [comment]
Looking forward to this, thanks for your work!

**Comment by p30arena** [comment]
can this pr solve this issue?
https://github.com/FirebaseExtended/flutterfire/issues/357

**Comment by tobiasjunsten** [comment]
@p30arena No, it only deals with handling push messages when app is in background or suspended state for iOS.

**Comment by campioncino** [comment]
Can I finnally see this beauviour in iOS? 
7bd39177-e115-4f8f-8aac-a035bcc808f1

**Comment by tobiasjunsten** [comment]
@campioncino this pull request is not adding that behaviour. It's just about adding background handling for push messages. So when your app is in the background you'll be able to provide a dart function/isolate that should handle the message.

**Comment by campioncino** [comment]
> @campioncino this pull request is not adding that behaviour. It's just about adding background handling for push messages. So when your app is in the background you'll be able to provide a dart function/isolate that should handle the message.

I'm sorry, I mean : can I finally do something like the image.

Using      `onBackgroundMessage: backgroundMessageHandler`
can I do something when my app is running in background?

**Comment by tobiasjunsten** [comment]
@campioncino Ok :) I guess that should be possible. Anything should be possible in the backgroundMessageHandler. Like fetching data from backend or trigger local notifications.

**Comment by campioncino** [comment]
> @campioncino Ok :) I guess that should be possible. Anything should be possible in the backgroundMessageHandler. Like fetching data from backend or trigger local notifications.

I hope so, but i can't handle it. I can't do anythings, either a simple print. Are you able to do that? Maybe I am doing something wrong.

**Comment by tobiasjunsten** [comment]
@campioncino I'm currently fetching data from backend and scheduling local notifications based on the data fetched from backend and that works fine.

There are some debug prints in the ios code. Could you please try to connect your phone to the computer and open the console app on the computer. Then select your phone and filter the console output by "didReceiveRemoteNotification". Then try to send a push notification when your app is open and see if you get any logs. If you get the logs, try to just hit the home button to get the app in the background and send a new push notification and see if you still get a log entry.

**Comment by campioncino** [comment]
IT WORKS!!!!

**Comment by misaelriojasftf** [comment]
how can I test this?

**Comment by tobiasjunsten** [comment]
@misaelriojasftf You can test this pullrequest by adding this to your pubspec.yaml:
```
firebase_messaging:
    git:
      url: git://github.com/abilia/flutterfire.git
      ref: ios-background-support
      path: packages/firebase_messaging/
```

**Comment by d3xt3r2909** [comment]
Firebase guys, can we speed a little bit this PR. 

This is quite an important and crucial feature for every app, cannot understand why it is not yet ready for use. Why do I need messages if they will not work in background !?

**Comment by kalismeras61** [comment]
can someone help me for using this? doesnt work for me?

**Comment by kalismeras61** [comment]
> IT WORKS!!!!

How ? any example ?

**Comment by kalismeras61** [comment]
When add AppDele

> @misaelriojasftf You can test this pullrequest by adding this to your pubspec.yaml:
> 
> ```
> firebase_messaging:
>     git:
>       url: git://github.com/abilia/flutterfire.git
>       ref: ios-background-support
>       path: packages/firebase_messaging/
> ```

I installed but it is not working. Is there any working repo ? for looking that ?

**Comment by kalismeras61** [comment]
> how can I test this?
@misaelriojasftf  did you test it ?

**Comment by kalismeras61** [comment]
> IT WORKS!!!!

@campioncino  any example ? or hint ?

**Comment by misaelriojasftf** [comment]
Hello @kalismeras61 , not unfortunately I couldn't. I just continue with other features. Push notifications are working so that's enough

**Comment by timosturm** [comment]
Does the background handler work for data messages also?

I try to show custom notifications and save incoming data from data-notifications, but the messages are still only received when the app is open or queried and received when the app is opened. I.e. only the callback for `onMessage` gets notifications.

I would appreciate any help.

@tobiasjunsten

**Comment by kalismeras61** [comment]
> Does the background handler work for data messages also?
> 
> I try to show custom notifications and save incoming data from data-notifications, but the messages are still only received when the app is open or queried and received when the app is opened. I.e. only the callback for `onMessage` gets notifications.
> 
> I would appreciate any help.
> 
> @tobiasjunsten

Yes it is working.

**Comment by jengchtk** [comment]
> @misaelriojasftf You can test this pullrequest by adding this to your pubspec.yaml:
> 
> ```
> firebase_messaging:
>     git:
>       url: git://github.com/abilia/flutterfire.git
>       ref: ios-background-support
>       path: packages/firebase_messaging/
> ```

How to use this PR. 
`Error output from Xcode build:
↳
    ** BUILD FAILED **
Xcode's output:
↳
    /Users/myaccount/.pub-cache/git/flutterfire-afcb41b3f0708be9eebe8bc405a0ed5cf63b14ad/packages/firebase_messaging/ios/Classes/FLTFirebaseMessagingPlugin.m:249:43: warning: 'FIRMessagingRemoteMessage' is deprecated: FCM direct channel is deprecated, please use APNs for downstream message handling. [-Wdeprecated-declarations]
    - (void)applicationReceivedRemoteMessage:(FIRMessagingRemoteMessage *)remoteMessage {
                                              ^
    In module 'FirebaseMessaging' imported from /Users/myaccount/my_app/ios/Pods/Headers/Public/Firebase/Firebase.h:72:
    /Users/myaccount/my_app/ios/Pods/FirebaseMessaging/Firebase/Messaging/Public/FIRMessaging.h:219:1: note: 'FIRMessagingRemoteMessage' has been explicitly marked deprecated here
    __deprecated_msg(
    ^
    In module 'UIKit' imported from /Users/myaccount/my_app/ios/Pods/Target Support Files/firebase_messaging/firebase_messaging-prefix.pch:2:
    In module 'Foundation' imported from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.5.sdk/System/Library/Frameworks/UIKit.framework/Headers/UIKit.h:8:
    In module 'CoreFoundation' imported from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.5.sdk/System/Library/Frameworks/Foundation.framework/Headers/Foundation.h:6:
    In module 'Darwin' imported from /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.5.sdk/System/Library/Frameworks/CoreFoundation.framework/Headers/CoreFoundation.h:16:
    /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator13.5.sdk/usr/include/sys/cdefs.h:191:48: note: expanded from macro '__deprecated_msg'
            #define __deprecated_msg(_msg) __attribute__((__deprecated__(_msg)))
                                                          ^`

**Comment by tobiasjunsten** [comment]
@jengchtk will look into that.

File: packages/firebase_messaging/README.md [hunk 1]
```diff
@@ -110,6 +148,15 @@ By default background messaging is not enabled. To handle messages in the backgr
    <application android:name=".Application" ...>
    ```
 
+#### iOS configuration (Swift)
+1. In `AppDelegate.swift`, add the following:
```
Comment (isaiahtaylor) [inline packages/firebase_messaging/README.md:152]: Note: `import firebase_messaging` needs to be added to the top of `AppDelegate.swift`. Worth mentioning.

**Comment by isaiahtaylor** [comment]
Thanks so much for your work on this. I was able to get it working as instructed. I did not get the deprecation error mentioned above. Let's get this merged!

**Comment by isaiahtaylor** [comment]
Hi, could you advise how it would be possible to use plugins from within the static handler method? I get `I/flutter ( 9944): MissingPluginException(No implementation found for method getAll on channel plugins.flutter.io/shared_preferences)` when I try to use `SharedPreferences` from within that function.

**Comment by michaelgobbers** [comment]
This sounds great however I've not been able to get this to work...
I've modified my pubspec as the comment above and made the changes needed to my appdelegate.swift. 

This is the message I'm sending to fcm
```curl -X POST --header "Authorization: key=<MY_KEY>" \
    --Header "Content-Type: application/json" \
    https://fcm.googleapis.com/fcm/send \
    -d "{\"to\":\"fQ08m1akPULFqLDlNl4u_o:APA91bGjzqSuLdMlkPs-xSwPY3j-d44991uktwsPHIvuWu94M62ZerLwscik9adCfKGj0kRmsinzFaEIrr_Q88ELgtH3wX_b75jom3loSlO-4dsW8ZzpePofYJ_ZNxIN0fCSRx_qpJtG\",\"data\":{\"yello\":\"Yellow\"},\"priority\":10}"
```

I basically only get the onMessage method to be triggered. Whenever I press the home button and trigger this nothing happens, when I then reopen, the regular onmessage is triggered again.
I am using the LocalNotificationsPlugin to display the silent notifications I want.

**Comment by kalismeras61** [comment]
> This sounds great however I've not been able to get this to work...
> I've modified my pubspec as the comment above and made the changes needed to my appdelegate.swift.
> 
> This is the message I'm sending to fcm
> 
> ```
> curl -X POST --header "Authorization: key=<MY_KEY>" \
>     --Header "Content-Type: application/json" \
>     https://fcm.googleapis.com/fcm/send \
>     -d "{\"to\":\"fQ08m1akPULFqLDlNl4u_o:APA91bGjzqSuLdMlkPs-xSwPY3j-d44991uktwsPHIvuWu94M62ZerLwscik9adCfKGj0kRmsinzFaEIrr_Q88ELgtH3wX_b75jom3loSlO-4dsW8ZzpePofYJ_ZNxIN0fCSRx_qpJtG\",\"data\":{\"yello\":\"Yellow\"},\"priority\":10}"
> ```
> 
> I basically only get the onMessage method to trigger. Whenever I press the home button and trigger this from terminal, nothing happens. When I then reopen, the regular onmessage is triggered again.
> I am using the LocalNotificationsPlugin to display the silent notifications I want.

Add also  "content_available": true,  as a parameter.

**Comment by tobiasjunsten** [comment]
> FIRMessagingRemoteMessage

@jengchtk I don't see any build errors in the log you provided. I only see warnings. Probably an error further down. Maybe you didn't import firebase_messaging as @isaiahtaylor pointed out is missing in the readme.

File: packages/firebase_messaging/README.md (hunk 1 above)
Comment (tobiasjunsten) [reply packages/firebase_messaging/README.md:152]: Good point! Easy to miss. I'll add that.

**Comment by michaelgobbers** [comment]
> 
> Add also "content_available": true, as a parameter.

Thx for your reply however. I'm still unable to get this to work. 
I first added the `"content_available":true` parameter but when I printed the resulting message it was added in a notification object. When I sent it like this `"content-available":true` it was not longer included in the message. However both messages only got print when I returned the app to the foreground. And the function that was being triggered was the onMessage function.

**Comment by kalismeras61** [comment]
> > Add also "content_available": true, as a parameter.
> 
> Thx for your reply however, I'm still unable to get this to work.
> I first added the `"content_available":true` parameter but when I printed the resulting message it was added in a notification object. When I sent it like this `"content-available":true` it was not longer included in the message. However both messages only got print when I returned the app to the foreground. And the function that was being triggered was the onMessage function.
> 
> for completeness this is what I'm sending:
> 
> ```
> "{\"to\":\"f62qaqn57UnWmlIBNh53mE:APA91bHhfgSI6DU5k-ci32M56uC3NexuSCu7OXYC0dWR172DsxMxXXbJ7Qa7KB0DiEvLT4z8ZGA9v_ta3tZlS6eVx-7Rm4aDBexn-QEbzD2aJ34w7J92oEsMdhoHEHZ-tTVWp7G1nbHZ\",\"content-available\":true,\"data\":{\"yello\":\"Yellow\"}}"
> ```

Use postman for testing. 
{
 "to" :"yourTokenId",
 "content_available": true,
 "priority" : "high",
 "data": {
 	"example_data": "yourData"
 }
}

**Comment by michaelgobbers** [comment]
> 
> Use postman for testing.
> {
> "to" :"yourTokenId",
> "content_available": true,
> "priority" : "high",
> "data": {
> "example_data": "yourData"
> }
> }

Same result I'm afraid.

**Comment by kalismeras61** [comment]
> > Use postman for testing.
> > {
> > "to" :"yourTokenId",
> > "content_available": true,
> > "priority" : "high",
> > "data": {
> > "example_data": "yourData"
> > }
> > }
> 
> Same result I'm afraid.

Thats weird. Are you able to get notification messages?

**Comment by michaelgobbers** [comment]
> > > Use postman for testing.
> > > {
> > > "to" :"yourTokenId",
> > > "content_available": true,
> > > "priority" : "high",
> > > "data": {
> > > "example_data": "yourData"
> > > }
> > > }
> > 
> > 
> > Same result I'm afraid.
> 
> Thats weird. Are you able to get notification messages?

yes I am but only when I resume it through the onMessage Callback.

This is how I set it up:
```
_initialize() {
    _firebaseMessaging.configure(
      onMessage: onMessage,
      onBackgroundMessage: showNotification,
    );
    _firebaseMessaging.requestNotificationPermissions(IosNotificationSettings(
        sound: true, alert: true, badge: true, provisional: false));
}
```

and these are my callback functions (first one is defined as a top level function so above the class definition that has the initialise method.
```
Future<dynamic> showNotification(Map<String, dynamic> message) async {
  print("background: $message");
  var initializationSettingsAndroid =
      AndroidInitializationSettings('@mipmap/rydoo_logo');
  var initializationSettingsIOS = IOSInitializationSettings(
    requestAlertPermission: true,
    requestBadgePermission: true,
    requestSoundPermission: true,
    defaultPresentAlert: true,
    defaultPresentBadge: true,
    defaultPresentSound: true,
  );
  var initializationSettings = InitializationSettings(
      initializationSettingsAndroid, initializationSettingsIOS);
  await plugin.initialize(initializationSettings);
  var androidDetails = AndroidNotificationDetails("0", "name", "description",
      priority: Priority.Max, importance: Importance.Max);
  var platformChannelSpecifics = NotificationDetails(
      androidDetails,
      IOSNotificationDetails(
          presentAlert: true, presentBadge: true, presentSound: true));
  return plugin.show(
    0,
    "ocr",
    jsonEncode(message),
    platformChannelSpecifics,
  );
}
```
```
Future onMessage(Map<String, dynamic> message) {
    print("front: $message");
    _streamController.add(message);
    return Future.value(0);
  }
```

**Comment by michaelgobbers** [comment]
I've been investigating further... I got it to work now. (log statement in console app) However I only got it to work by disabling a bunch of other plugins. I'm trying to figure out which one is causing this behaviour.

**Comment by kalismeras61** [comment]
> I've been investigating further... I got it to work now. (log statement in console app) However I only got it to work by disabling a bunch of other plugins. I'm trying to figure out which one is causing this behaviour.

Okay cool. Please let me know i am also using this PR.

**Comment by Fatimamostafa** [comment]
> @misaelriojasftf You can test this pullrequest by adding this to your pubspec.yaml:
> 
> ```
> firebase_messaging:
>     git:
>       url: git://github.com/abilia/flutterfire.git
>       ref: ios-background-support
>       path: packages/firebase_messaging/
> ```

Is it supposed to work along with the existing `onBackgroundMessage: _backgroundMessageHandler`? Cause I've just added it to `pubspec.yaml` but it's not working? Anything else needs to be taken care of?

**Comment by tobiasjunsten** [comment]
> Hi, could you advise how it would be possible to use plugins from within the static handler method? I get `I/flutter ( 9944): MissingPluginException(No implementation found for method getAll on channel plugins.flutter.io/shared_preferences)` when I try to use `SharedPreferences` from within that function.
> 
> Edit: this seems to be related to [flutter/flutter#13937](https://github.com/flutter/flutter/issues/13937).

I'm using SharedPreferences in the background isolate so it's possible. I don't remember doing anything special to get it to work. Will try to see if I have some specific setting. I've had similar problems with other libraries though, like flutter_secure_storage.

**Comment by michaelgobbers** [comment]
> > I've been investigating further... I got it to work now. (log statement in console app) However I only got it to work by disabling a bunch of other plugins. I'm trying to figure out which one is causing this behaviour.
> 
> Okay cool. Please let me know i am also using this PR.

Alright so I've pinpointed the conflicting plugin. My project also contains a plugin for intercom [https://pub.dev/packages/intercom_flutter](url) It also offers push notifications.

**Comment by fanatic75** [comment]
I have tried this pull,  so when I add content_available to my payload, the app suddenly loses connection with the console and I cannot read logs anymore. 
`{
 "to" : "",
 
     "content_available":true,
 "data" : {
     
     "click_action": "FLUTTER_NOTIFICATION_CLICK",
     "title":"news",
     "body":"body",
     "newsID":200,
     "image_url":"https://media.wired.com/photos/5b899992404e112d2df1e94e/master/pass/trash2-01.jpg",
 }
}`

**Comment by isaiahtaylor** [comment]
Hey all, I'll share what we had to do on the plugin front to get this to work. I can't vouch for whether this will be relevant for everyone.

We initially had issues with SharedPreferences, getting this error:

```
I/flutter ( 9944): MissingPluginException(No implementation found for method getAll on channel plugins.flutter.io/shared_preferences)
```


```
❯ flutter --version
Flutter 1.12.13+hotfix.9 • channel unknown • unknown source
Framework • revision f139b11009 (3 months ago) • 2020-03-30 13:57:30 -0700
Engine • revision af51afceb8
Tools • Dart 2.7.2
```

In our `Application.kt`, in the `registerWith` function, we had to add several plugins to get them to work in that static function:

```kt
package com.[removed].[removed];

import io.flutter.app.FlutterApplication
import io.flutter.plugin.common.PluginRegistry
import io.flutter.plugin.common.PluginRegistry.PluginRegistrantCallback
import io.flutter.plugins.firebasemessaging.FirebaseMessagingPlugin
import io.flutter.view.FlutterMain
import io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService
import io.flutter.plugins.sharedpreferences.SharedPreferencesPlugin
import io.flutter.plugins.pathprovider.PathProviderPlugin
import io.flutter.plugins.firebaseanalytics.FirebaseAnalyticsPlugin

class Application : FlutterApplication(), PluginRegistrantCallback {

    override fun onCreate() {
        super.onCreate()
        FlutterFirebaseMessagingService.setPluginRegistrant(this);
        FlutterMain.startInitialization(this)
    }

    override fun registerWith(registry: PluginRegistry?) {
        if (!registry!!.hasPlugin("io.flutter.plugins.firebasemessaging")) {
            FirebaseMessagingPlugin.registerWith(registry!!.registrarFor("io.flutter.plugins.firebasemessaging.FirebaseMessagingPlugin"));
            SharedPreferencesPlugin.registerWith(registry!!.registrarFor("io.flutter.plugins.sharedpreferences.SharedPreferencesPlugin"));
            FirebaseAnalyticsPlugin.registerWith(registry!!.registrarFor("io.flutter.plugins.firebaseanalytics.FirebaseAnalyticsPlugin"));
        }
    }
}
```

Not sure if this would be different on newer version of Flutter.

**Comment by fanatic75** [comment]
@isaiahtaylor, please don't post android config on a thread for iOS. There is no point here.

**Comment by isaiahtaylor** [comment]
I edited my comment to note that this is not necessarily relevant to this issues, it's just something that you might run into following this path.

**Comment by tobiasjunsten** [comment]
> @tobiasjunsten , I have tried this pull, so when I add content_available to my payload, the app suddenly loses connection with the console and I cannot read logs anymore.
> `{
> "to" : "",
> 
> ```
>  "content_available":true,
> ```
> 
> "data" : {
> 
> ```
>  "click_action": "FLUTTER_NOTIFICATION_CLICK",
>  "title":"news",
>  "body":"body",
>  "newsID":200,
>  "image_url":"https://media.wired.com/photos/5b899992404e112d2df1e94e/master/pass/trash2-01.jpg",
> ```
> 
> }
> }`
> 
> If I remove the content_available:true, it doesn't do anything in the background with this branch. but as soon as I add, it loses connection with the machine and I cannot read further logs to see If I have received the data message

@fanatic75 have you tried the console app on iOS? If you lose the connection with the app in the IDE-console you should still be able to see the logs from the app in the stand alone console app.

**Comment by fanatic75** [comment]
@

> > @tobiasjunsten , I have tried this pull, so when I add content_available to my payload, the app suddenly loses connection with the console and I cannot read logs anymore.
> > `{
> > "to" : "",
> > ```
> >  "content_available":true,
> > ```
> > 
> > 
> > "data" : {
> > ```
> >  "click_action": "FLUTTER_NOTIFICATION_CLICK",
> >  "title":"news",
> >  "body":"body",
> >  "newsID":200,
> >  "image_url":"https://media.wired.com/photos/5b899992404e112d2df1e94e/master/pass/trash2-01.jpg",
> > ```
> > 
> > 
> > }
> > }`
> > If I remove the content_available:true, it doesn't do anything in the background with this branch. but as soon as I add, it loses connection with the machine and I cannot read further logs to see If I have received the data message
> 
> @fanatic75 have you tried the console app on iOS? If you lose the connection with the app in the IDE-console you should still be able to see the logs from the app in the stand alone console app.

I have moved from data message to notification message which solves the background problem on IOS as well. OnLaunch is working once the app has been killed and onResume is working when the app is in background. Not using background message callback anymore on both platforms. 
Though there is definitely something weird as the connection should not get closed and the app is still running fine without crashing.

File: packages/firebase_messaging/README.md [hunk 2]
```diff
@@ -207,12 +226,12 @@ Next, you should probably request permissions for receiving Push Notifications.
...
+|     **Notification on iOS** | `onMessage`       | Notification is delivered to system tray. When the user clicks on it to open app `onResume` fires.                                                                  | Notification is delivered to system tray. When the user clicks on it to open app `onLaunch` fires.                                                                  |
+| **Data Message on Android** | `onMessage`       | `onMessage` while app stays in the background.                                                                                                                      | *not supported by plugin, message is lost*                                                                                                                          |
+|     **Data Message on iOS** | `onMessage`       | Message is stored by FCM and delivered to app via `onMessage` when the app is brought back to foreground.                                                           | Message is stored by FCM and delivered to app via `onMessage` when the app is brought back to foreground.                                                           |
```
Comment (oskara) [inline packages/firebase_messaging/README.md:234]: Is this right? Aren't background messages delivered through `onBackgroundMessage` after this patch?

File: packages/firebase_messaging/README.md [hunk 3]
```diff
@@ -110,6 +148,22 @@ By default background messaging is not enabled. To handle messages in the backgr
...
    ```
 
+#### iOS configuration (Swift)
+1. In the top of `AppDelegate.swift`, add the import of firebase_messaging:
+   
+   ```swift
+    import firebase_messaging
+   ```
+
+1. Then add the following code to `AppDelegate.swift`:
```
Comment (oskara) [inline packages/firebase_messaging/README.md:158]: Maybe you should specify where in `AppDelegate.swift`? E.g. `didFinishLaunchingWithOptions` if that is the case.

**Comment by SKKurz** [comment]
is it also possible to use the background message handling on an iOS Simulator? Or is this one of the things that only work on a proper iOS device?

**Comment by oskara** [comment]
> is it also possible to use the background message handling on an iOS Simulator? Or is this one of the things that only work on a proper iOS device?

As far as I know this will only work on real devices.

**Comment by oskara** [comment]
I was able to get this to work on iOS12, but after upgrading to 13 it didn't work any more. Did anyone try this on iOS13 and above?

I'm getting:
```
NO LONGER RUNNING [Not in application -> activity mapping] com.apple.pushLaunch.com.myapp.myapp:ABC5110 ...Tasks running in group [com.apple.dasd.defaultNetwork] are 1!
```
Anyone else seen this?

**Comment by owenkealey** [comment]
Only conflicts are changelog and pubspec? This is a critical merge for my app.

**Comment by csarigumba** [comment]
Is this is the correct way writing this in Objective-C?

```
void registerPlugins(NSObject<FlutterPluginRegistry>* registry) {
  [GeneratedPluginRegistrant registerWithRegistry:registry];
}

[FLTFirebaseMessagingPlugin setPluginRegistrantCallback:registerPlugins];
```

Please advise.

File: packages/firebase_messaging/README.md [hunk 4]
```diff
@@ -205,14 +224,14 @@ Next, you should probably request permissions for receiving Push Notifications.
...
+|     **Notification on iOS** | `onMessage`       | Notification is delivered to system tray. When the user clicks on it to open app `onResume` fires.                                                                  | Notification is delivered to system tray. When the user clicks on it to open app `onLaunch` fires.                                                                  |
+| **Data Message on Android** | `onMessage`       | `onMessage` while app stays in the background.                                                                                                                      | *not supported by plugin, message is lost*                                                                                                                          |
+|     **Data Message on iOS** | `onMessage`       | Message is delivered to `onBackgroundMessage`. This is the case both when app is running in background and the system has suspended the app.                        | Whan app is force-quit by the user the message is not handled.                                                                                                      |
```
[the remark this answers is not in the document]
Comment (tobiasjunsten) [reply packages/firebase_messaging/README.md]: Thanks. Fixed.

**Comment by tobiasjunsten** [comment]
Hi @FirebaseExtended/invertase! Any chance you could take a look at this pull request? It's a really wanted feature for us and (as it seems) a lot of other people.

**Comment by tobiasjunsten** [comment]
Hi @FirebaseExtended/invertase! Any chance you could take a look at this pull request? It's a really wanted feature for us and (as it seems) a lot of other people.

**Comment by tobiasjunsten** [comment]
Hi @Salakar! I tried to tag the @FirebaseExtended/invertase team as suggested here: [FlutterFire Roadmap](https://github.com/FirebaseExtended/flutterfire/issues/2582) but don't seem to work. Any chance you could take a look at this pull request?

**Comment by kristijandraca** [comment]
Any updates on this?

**Comment by nt4f04uNd** [comment]
@collinjackson @Ehesp @kroikie @Salakar guys please review this one, we all need it 🙏🥺

**Comment by Ehesp** [comment]
Hold on to your hats... New dev release coming soon.

**Comment by ozcannn14** [comment]
When i send data notification like ` data: { "click_action": "FLUTTER_NOTIFICATION_CLICK", "status": "done", "notification": JSON.stringify(notification1) },` and when app is foreground the app is crashing in IOS platform.

> 
> 2020-10-29 16:51:57.183699+0300 Runner[23312:110449] -[__NSCFString setObject:forKeyedSubscript:]: unrecognized selector sent to instance 0x6000029faa00
> 2020-10-29 16:51:57.223837+0300 Runner[23312:110449] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSCFString setObject:forKeyedSubscript:]: unrecognized selector sent to instance 0x6000029faa00'
> *** First throw call stack:
> (
> 	0   CoreFoundation                      0x00007fff2043a126 __exceptionPreprocess + 242
> 	1   libobjc.A.dylib                     0x00007fff20177f78 objc_exception_throw + 48
> 	2   CoreFoundation                      0x00007fff20448c6f +[NSObject(NSObject) instanceMethodSignatureForSelector:] + 0
> 	3   CoreFoundation                      0x00007fff2043e666 ___forwarding___ + 1489
> 	4   CoreFoundation                      0x00007fff20440698 _CF_forwarding_prep_0 + 120
> 	5   Runner                              0x0000000107e248bd -[FIRMessagingDataMessageManager parseDataMessage:] + 1565
> 	6   Runner                              0x0000000107e241af -[FIRMessagingDataMessageManager processPacket:] + 335
> 	7   Runner                              0x0000000107e1b29e -[FIRMessagingClient connectionDidRecieveMessage:] + 94
> 	8   Runner                              0x0000000107e2024b -[FIRMessagingConnection didReceiveDataMessageStanza:] + 91
> 	9   Runner                              0x0000000107e1e851 -[FIRMessagingConnection secureSocket:didReceiveData:withTag:] + 801
> 	10  Runner                              0x0000000107e3bdd7 -[FIRMessagingSecureSocket processCurrentInputBuffer:outOffset:] + 583
> 	11  Runner                              0x0000000107e3b914 -[FIRMessagingSecureSocket performRead] + 2276
> 	12  Runner                              0x0000000107e3a77f -[FIRMessagingSecureSocket stream:handleEvent:] + 239
> 	13  CoreFoundation                      0x00007fff203cb29c _signalEventSync + 217
> 	14  CoreFoundation                      0x00007fff203c89f0 _cfstream_solo_signalEventSync + 266
> 	15  CoreFoundation                      0x00007fff203c88a8 _CFStreamSignalEvent + 479
> 	16  CFNetwork                           0x00007fff235ea914 _CFStreamErrorFromCFError + 198638
> 	17  CFNetwork                           0x00007fff235f1f93 _CFStreamErrorFromCFError + 228973
> 	18  CoreFoundation                      0x00007fff203bb739 __CFSocketPerformV0 + 932
> 	19  CoreFoundation                      0x00007fff203a8845 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
> 	20  CoreFoundation                      0x00007fff203a873d __CFRunLoopDoSource0 + 180
> 	21  CoreFoundation                      0x00007fff203a7c1f __CFRunLoopDoSources0 + 248
> 	22  CoreFoundation                      0x00007fff203a23f7 __CFRunLoopRun + 878
> 	23  CoreFoundation                      0x00007fff203a1b9e CFRunLoopRunSpecific + 567
> 	24  GraphicsServices                    0x00007fff2b773db3 GSEventRunModal + 139
> 	25  UIKitCore                           0x00007fff24660af3 -[UIApplication _run] + 912
> 	26  UIKitCore                           0x00007fff24665a04 UIApplicationMain + 101
> 	27  Runner                              0x00000001078ae9eb main + 75
> 	28  libdyld.dylib                       0x00007fff20257415 start + 1
> )
> libc++abi.dylib: terminating with uncaught exception of type NSException
> *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[__NSCFString setObject:forKeyedSubscript:]: unrecognized selector sent to instance 0x6000029faa00'
> terminating with uncaught exception of type NSException
> CoreSimulator 732.17 - Device: iPhone 11 Pro Max (5BA6EABD-B326-4850-B830-30F3DA855DC4) - Runtime: iOS 14.0 (18A372) - DeviceType: iPhone 11 Pro Max

**Comment by nt4f04uNd** [comment]
@ozcannn14 why would you JSON.stringify your `"notification"`?

**Comment by ozcannn14** [comment]
> @ozcannn14 why would you JSON.stringify your `"notification"`? If you are sending this with the JS SDK it has to be an object I guess

@nt4f04uNd because i want to trigger fcm background handle method and run some process then show notification with the local notification plugin.

**Comment by nt4f04uNd** [comment]
@ozcannn14 then use data messages instead https://firebase.google.com/docs/cloud-messaging/concept-options

**Comment by ozcannn14** [comment]
> @ozcannn14 then use data messages instead https://firebase.google.com/docs/cloud-messaging/concept-options

@nt4f04uNd İt is not a notification this is just a tag :)

**Comment by gahms** [comment]
I have used this PR and it works. One thing that bit me was that when testing with a local clone of the fork, it seems that firebase_messaging gets initialized before firebase_core and that is not handled very well in the `FLTFirebaseMessagingPlugin initWithChannel:registrar:]` method that fails silently (`[FIRMessaging messaging]` is `nil`).

An entirely different issue is that because of https://github.com/flutter/flutter/issues/14815, I cannot do any of the processing I had in mind in the background isolate and thus it is effectively useless to me to have background support on iOS.

**Comment by nt4f04uNd** [comment]
Hey, I'm trying to use the pull request, but the app crashes with the following crashlog. Help wanted asap
[crashlog.log](https://github.com/FirebaseExtended/flutterfire/files/5477920/crashlog.log)

**Pull request closed without merging**